File: my-folder/0022-generate-parentheses/solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        recursion(result, "", 0, 0, n);
        return result;
    }

    private void recursion(List<String> result, String current, int open, int close, int n) {
        if (current.length() == 2 * n) {
            result.add(current);
            return;
        }
        if (open < n) {
            recursion(result, current + '(', open + 1, close, n);
        }
        if (close < open) {
            recursion(result, current + ')', open, close + 1, n);
        }
    }
}

File: my-folder/0022-generate-parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # Backtracking solution
        stack = []
        result = []

        def backtrack(openN, closeN):
            if openN == closeN == n:
                result.append("".join(stack))
                return result

            if openN < n:
                stack.append("(")
                backtrack(openN + 1, closeN)
                stack.pop()

            if closeN < openN:
                stack.append(")")
                backtrack(openN, closeN + 1)
                stack.pop()

        backtrack(0, 0)
        return result

# Time complexity: O(4^n/sqrt(n)) (Catalan number)
# Space complexity: O(n)
